### DISCUSSION TITLE
what is the best library to create customisable forms in MDX ?

### DISCUSSION THREAD

**Discussion by sandys** [mdx-js/mdx#1738] [Q&A] [answered]
hi,
so this question is from a developer experience (DX) perspective. I know that the technical way to create a form in react is to use form components. 

Want to check if anybody in the community has done this and has any learnings here ? I was thinking of using Formik, but im unsure if the end-user experience is all that great. Remember if someone is using MDX to do stuff...then we already have a person who is not super technical anyway.

Which libraries have you seen having the best adoption from end-user DX perspective ? it may not be fully featured, but i want to have the markdown writing experience as intuitive and ...well...close to markdown/text as possible.

**Comment by ChristianMurphy** [comment, marked as answer]
To preface, it's worth noting that software is very contextual.
Why a library or solution is needed influences what an appropriate decision would be.
And the existing technologies and culture of an organization influence decisions as well.
The isn't a clear-cut and one-size-fits-all answer to what is "best" or what should be done.

For example in your question.
Why/how are forms being used? Who will be authoring the forms? What front-end technology (React, Vue, Preact, Svelte)? And what design system (Bootstrap, Material Design, custom, etc) is being used?
All of these would significantly change the answer to the question.

Instead what I can offer are some considerations on what to look for and consider when making the decision.

## Do I need a component library?

JSX supports plain HTML elements like `form` and `input`, which can be used to post information to a server.
In many cases leveraging plain HTML and HTTP is easier.

And even if client side handling of forms is needed, understand what a given form library is bringing.
In many cases what is generally wanted is a pattern (for example https://twitter.com/housecor/status/1437765736026542081)

## What are the MDX authors familiar with?

Consider the people who will be creating the content.
From a syntax perspective, are author more familiar with JSX or with [markdown directives](https://github.com/remarkjs/remark-directive)?
If JSX, have they created forms before? and using what technologies? Did those technologies work well for them? Could they be re-used?

## What technologies are being used in my stack?

Content usually isn't stand alone, it is part of a larger product/ecosystem.
What other technologies are in use? Are those technologies working well? Could the be used in the MDX content as well?

## What style-guide or design-system is used?

Are material design, bootstrap, another design system, or a home-grown/custom design system being used?
Are there components from the design system that could be re-used?

## How is the library and it's community doing?

Use comparison tools like: https://moiva.io, https://www.npmtrends.com, and https://bundlephobia.com to look at a package, relative to similar packages.
Using your preferred metrics for comparison.
Some popular metrics to look at:
* Community: how many people are contributing, how many people are adopting, how is that changing
* Quality: using metrics like complexity measures, security measure, performance/size measures
* Stability: how often do releases happen? How many are breaking changes? How long has a package been around

## If the library has to be replaced, how much would need to change?

Sometimes libraries stop being maintained.
Consider if that happens, how easy or hard would it be to migrate to another library?
Is there an alternative library with a similar or identical API?
